chain/bitcoind_client.go

```diff
@@ -36,99 +32,90 @@ type BitcoindClient struct {
+	notificationQueue *ConcurrentQueue
 
-	memPool    map[chainhash.Hash]struct{}
-	memPoolExp map[int32]map[chainhash.Hash]struct{}
+	// zmqTxNtfns is a channel through which ZMQ transaction events will be
+	// retrieved from the backing bitcoind connection.
+	zmqTxNtfns chan *wire.MsgTx
+
+	// zmqBlockNtfns is a channel through which ZMQ block events will be
+	// retrieved from the backing bitcoind connection.
+	zmqBlockNtfns chan *wire.MsgBlock
 
 	quit chan struct{}
 	wg   sync.WaitGroup
 }
 
-// NewBitcoindClient creates a client connection to the server described by the
-// connect string.  If disableTLS is false, the remote RPC certificate must be
-// provided in the certs slice.  The connection is not established immediately,
-// but must be done using the Start method.  If the remote server does not
-// operate on the same bitcoin network as described by the passed chain
-// parameters, the connection will be disconnected.
-func NewBitcoindClient(chainParams *chaincfg.Params, connect, user, pass,
-	zmqConnect string, zmqPollInterval time.Duration) (*BitcoindClient,
-	error) {
-
-	client := &BitcoindClient{
-		connConfig: &rpcclient.ConnConfig{
-			Host:                 connect,
-			User:                 user,
-			Pass:                 pass,
-			DisableAutoReconnect: false,
-			DisableConnectOnNew:  true,
-			DisableTLS:           true,
-			HTTPPostMode:         true,
-		},
-		chainParams:     chainParams,
-		zmqConnect:      zmqConnect,
-		zmqPollInterval: zmqPollInterval,
-		rescanUpdate:    make(chan interface{}),
-		watchOutPoints:  make(map[wire.OutPoint]struct{}),
-		watchAddrs:      make(map[string]struct{}),
-		watchTxIDs:      make(map[chainhash.Hash]struct{}),
-		quit:            make(chan struct{}),
-		memPool:         make(map[chainhash.Hash]struct{}),
-		memPoolExp:      make(map[int32]map[chainhash.Hash]struct{}),
-	}
-	rpcClient, err := rpcclient.New(client.connConfig, nil)
-	if err != nil {
-		return nil, err
-	}
-	client.client = rpcClient
-	return client, nil
-}
-
 // BackEnd returns the name of the driver.
 func (c *BitcoindClient) BackEnd() string {
 	return "bitcoind"
 }
 
-// GetCurrentNet returns the network on which the bitcoind instance is running.
-func (c *BitcoindClient) GetCurrentNet() (wire.BitcoinNet, error) {
-	hash, err := c.client.GetBlockHash(0)
-	if err != nil {
-		return 0, err
-	}
-
-	switch *hash {
-	case *chaincfg.TestNet3Params.GenesisHash:
-		return chaincfg.TestNet3Params.Net, nil
-	case *chaincfg.RegressionNetParams.GenesisHash:
-		return chaincfg.RegressionNetParams.Net, nil
-	case *chaincfg.MainNetParams.GenesisHash:
-		return chaincfg.MainNetParams.Net, nil
-	default:
-		return 0, errors.New("unknown network")
-	}
-}
-
 // GetBestBlock returns the highest block known to bitcoind.
 func (c *BitcoindClient) GetBestBlock() (*chainhash.Hash, int32, error) {
-	bcinfo, err := c.client.GetBlockChainInfo()
+	bcinfo, err := c.chainConn.client.GetBlockChainInfo()
```

Comment:
I think it is actually better to return the `bestBlock` here, as we might risk that and what gets returned from the backend to be out of sync, and I just imagine that could lead to weird issues.